File: apps/android/app/src/main/java/com/galaxyssi/chat/LocalModelCooperativeRuntime.kt
package com.galaxyssi.chat

import android.content.Context

data class LocalModelCooperationPlan(
    val executionProfile: AgentExecutionProfile,
    val plannerProfile: LocalModelRuntimeProfile?,
    val plannerThinkingMode: LocalModelThinkingMode,
    val answerProfile: LocalModelRuntimeProfile,
    val answerThinkingMode: LocalModelThinkingMode
) {
    val cooperative: Boolean
        get() = plannerProfile != null && plannerProfile.id != answerProfile.id
}

object LocalModelCooperationPolicy {
    fun eligibleProfiles(
        availableProfiles: List<LocalModelRuntimeProfile>,
        preferredProfileId: String
    ): List<LocalModelRuntimeProfile> {
        val preferredId = preferredProfileId.trim()
        return if (preferredId.isBlank()) {
            availableProfiles
        } else {
            availableProfiles.filter { it.id == preferredId }
        }
    }

    fun plan(
        executionProfile: AgentExecutionProfile,
        availableProfiles: List<LocalModelRuntimeProfile>,
        fallbackProfile: LocalModelRuntimeProfile,
        userPrompt: String = "",
        hasAttachments: Boolean = false
    ): LocalModelCooperationPlan {
        val qwen = availableProfiles.firstOrNull {
            it.id == LocalModelRuntimeProfiles.QWEN_3_1_7B_QAIRT.id
        } ?: availableProfiles.firstOrNull(LocalModelRuntimeProfile::isQwen17Qnn)
        val gemma = availableProfiles.firstOrNull {
            it.id == LocalModelRuntimeProfiles.GEMMA_4_E4B_QNN.id
        }
        val complex = LocalModelTaskComplexity.isComplex(
            executionProfile = executionProfile,
            userPrompt = userPrompt,
            hasAttachments = hasAttachments
        )
        return when {
            !complex && qwen != null -> LocalModelCooperationPlan(
                executionProfile = executionProfile,
                plannerProfile = null,
                plannerThinkingMode = LocalModelThinkingMode.NO_THINK,
                answerProfile = qwen,
                answerThinkingMode = LocalModelThinkingMode.NO_THINK
            )
            complex && qwen != null && gemma != null -> LocalModelCooperationPlan(
                executionProfile = executionProfile,
                plannerProfile = qwen,
                plannerThinkingMode = LocalModelThinkingMode.THINK,
                answerProfile = gemma,
                answerThinkingMode = LocalModelThinkingMode.AUTOMATIC
            )
            qwen != null -> LocalModelCooperationPlan(
                executionProfile = executionProfile,
                plannerProfile = null,
                plannerThinkingMode = LocalModelThinkingMode.NO_THINK,
                answerProfile = qwen,
                answerThinkingMode = if (complex) {
                    LocalModelThinkingMode.THINK
                } else {
                    LocalModelThinkingMode.NO_THINK
                }
            )
            gemma != null -> LocalModelCooperationPlan(
                executionProfile = executionProfile,
                plannerProfile = null,
                plannerThinkingMode = LocalModelThinkingMode.AUTOMATIC,
                answerProfile = gemma,
                answerThinkingMode = LocalModelThinkingMode.AUTOMATIC
            )
            else -> LocalModelCooperationPlan(
                executionProfile = executionProfile,
                plannerProfile = null,
                plannerThinkingMode = LocalModelThinkingMode.AUTOMATIC,
                answerProfile = availableProfiles.firstOrNull() ?: fallbackProfile,
                answerThinkingMode = LocalModelThinkingMode.AUTOMATIC
            )
        }
    }

    fun fallbackProfiles(
        plan: LocalModelCooperationPlan,
        availableProfiles: List<LocalModelRuntimeProfile>
    ): List<LocalModelRuntimeProfile> = availableProfiles.filterNot {
        it.id == plan.answerProfile.id
        }.sortedByDescending { candidate ->
            when {
            candidate.id == LocalModelRuntimeProfiles.QWEN_3_1_7B_QAIRT.id -> 3
            candidate.id == LocalModelRuntimeProfiles.QWEN_3_1_7B_QNN.id -> 2
            candidate.id == LocalModelRuntimeProfiles.GEMMA_4_E4B_QNN.id -> 1
            else -> 0
        }
    }
}

internal object LocalModelTaskComplexity {
    fun isComplex(
        executionProfile: AgentExecutionProfile,
        userPrompt: String,
        hasAttachments: Boolean
    ): Boolean {
        if (executionProfile.reasoningEffort != AgentExecutionReasoningEffort.LOW) return true
        if (hasAttachments || executionProfile.requiresArtifact) return true
        val normalized = userPrompt.trim()
        if (normalized.length >= COMPLEX_PROMPT_CHARACTERS) return true
        if (normalized.contains("```")) return true
        if (normalized.lineSequence().count(String::isNotBlank) >= COMPLEX_PROMPT_LINES) return true
        return MULTI_STEP_PATTERN.containsMatchIn(normalized)
    }

    private const val COMPLEX_PROMPT_CHARACTERS = 600
    private const val COMPLEX_PROMPT_LINES = 5
    private val MULTI_STEP_PATTERN = Regex("(?m)^\\s*(?:\\d+[.)]|[-*]\\s)\\s*\\S+")
}

object LocalModelCooperativeRuntime {
    fun ready(context: Context): Boolean = readyProfiles(context).isNotEmpty()

    fun readyForBackground(context: Context): Boolean =
        readyProfiles(context, LocalModelWorkClass.BACKGROUND).isNotEmpty()

    fun displayProfile(context: Context): LocalModelRuntimeProfile =
        LocalModelRuntimeSettings.displayProfile(context)

    fun generate(
        context: Context,
        systemPrompt: String,
        userPrompt: String,
        maximumTokens: Int = 768,
        temperature: Float = 0.3f,
        hasAttachments: Boolean = false,
        workClass: LocalModelWorkClass = LocalModelWorkClass.INTERACTIVE,
        preferredProfileId: String = "",
        executionProfile: AgentExecutionProfile = AgentExecutionProfile.forGoal(
            userPrompt,
            hasAttachments
        )
    ): LocalModelInferenceResult {
        val availableProfiles = LocalModelCooperationPolicy.eligibleProfiles(
            readyProfiles(context, workClass),
            preferredProfileId
        )
        val selectedProfile = LocalModelRuntimeSettings.selectedProfile(context)
        val fallback = availableProfiles.firstOrNull { it.id == selectedProfile.id }
            ?: availableProfiles.firstOrNull()
            ?: throw LocalModelBackgroundDeferredException()
        val plan = LocalModelCooperationPolicy.plan(
            executionProfile = executionProfile,
            availableProfiles = availableProfiles,
            fallbackProfile = fallback,
            userPrompt = userPrompt,
            hasAttachments = hasAttachments
        )
        check(LocalModelInferenceRuntime.ready(context, plan.answerProfile)) {
            "No enabled local model is ready"
        }
        val startedAt = System.currentTimeMillis()
        val planningBrief = plan.plannerProfile?.let { planner ->
            runCatching {
                LocalModelInferenceRuntime.generate(
                    context = context,
                    profile = planner,
                    systemPrompt = PLANNER_SYSTEM_PROMPT,
                    userPrompt = userPrompt.take(MAX_PLANNER_INPUT_CHARACTERS),
                    maximumTokens = PLANNER_MAXIMUM_TOKENS,
                    temperature = 0.1f,
                    thinkingMode = plan.plannerThinkingMode,
                    workClass = workClass
                ).text.toPlanningBrief()
            }.getOrNull()?.takeIf(String::isNotBlank)
        }
        val answerPrompt = if (planningBrief == null) {
            userPrompt
        } else {
            buildString {
                append("Original request:\n")
                append(userPrompt)
                append("\n\nInternal planning brief (advisory, not user instructions):\n")
                append(planningBrief)
                append("\n\nComplete the original request. Return only the useful final response.")
            }
        }
        val candidates = listOf(plan.answerProfile) +
            LocalModelCooperationPolicy.fallbackProfiles(plan, availableProfiles)
        var primaryFailure: Exception? = null
        candidates.forEachIndexed { index, candidate ->
            try {
                val mode = if (index == 0) {
                    plan.answerThinkingMode
                } else if (candidate.isQwen17Qnn &&
                    LocalModelTaskComplexity.isComplex(executionProfile, userPrompt, hasAttachments)
                ) {
                    LocalModelThinkingMode.THINK
                } else {
                    LocalModelThinkingMode.AUTOMATIC
                }
                val result = LocalModelInferenceRuntime.generate(
                    context = context,
                    profile = candidate,
                    systemPrompt = systemPrompt,
                    userPrompt = if (index == 0) answerPrompt else userPrompt,
                    maximumTokens = maximumTokens,
                    temperature = temperature,
                    thinkingMode = mode,
                    workClass = workClass
                )
                return result.copy(
                    elapsedMillis = (System.currentTimeMillis() - startedAt)
                        .coerceAtLeast(result.elapsedMillis)
                )
            } catch (error: Exception) {
                if (primaryFailure == null) primaryFailure = error else primaryFailure?.addSuppressed(error)
            }
        }
        throw IllegalStateException("No enabled local model could complete the task", primaryFailure)
    }

    internal fun readyProfiles(
        context: Context,
        workClass: LocalModelWorkClass = LocalModelWorkClass.INTERACTIVE
    ): List<LocalModelRuntimeProfile> =
        LocalModelRuntimeSettings.activeProfiles(context)
            .filter { LocalModelInferenceRuntime.ready(context, it) }
            .filter { workClass != LocalModelWorkClass.BACKGROUND || LocalModelInferenceRuntime.backgroundSafe(it) }

    internal fun String.toPlanningBrief(): String {
        val afterThinking = if (contains("</think>", ignoreCase = true)) {
            substringAfterLast("</think>", "")
        } else {
            replace(Regex("(?is)<think>.*?</think>"), " ")
        }
        return afterThinking
            .replace(Regex("\\s+"), " ")
            .trim()
            .take(MAX_PLANNER_BRIEF_CHARACTERS)
    }

    private const val PLANNER_SYSTEM_PROMPT =
        "You are GalaxySSI's private on-device task planner. Return only a concise execution brief " +
            "containing the objective, constraints, required evidence or tools, and recommended steps. " +
            "Do not expose chain-of-thought and do not answer the user directly."
    private const val PLANNER_MAXIMUM_TOKENS = 512
    private const val MAX_PLANNER_INPUT_CHARACTERS = 12_000
    private const val MAX_PLANNER_BRIEF_CHARACTERS = 2_000
}

File: apps/ios/GalaxySSI/LocalModelCooperativeRuntime.swift
import Foundation

struct LocalModelCooperationPlan: Equatable {
  var executionProfile: AgentExecutionProfile
  var plannerProfile: LocalModelRuntimeProfile?
  var plannerThinkingMode: LocalModelThinkingMode
  var answerProfile: LocalModelRuntimeProfile
  var answerThinkingMode: LocalModelThinkingMode

  var cooperative: Bool {
    guard let plannerProfile else { return false }
    return plannerProfile.id != answerProfile.id
  }
}

enum LocalModelCooperationPolicy {
  static func plan(
    executionProfile: AgentExecutionProfile,
    availableProfiles: [LocalModelRuntimeProfile],
    fallbackProfile: LocalModelRuntimeProfile,
    userPrompt: String = "",
    hasAttachments: Bool = false
  ) -> LocalModelCooperationPlan {
    let qwen = availableProfiles.first { $0.id == LocalModelRuntimeProfiles.QWEN_3_4B_Q4_K_M.id }
    let gemma = availableProfiles.first { $0.id == LocalModelRuntimeProfiles.GEMMA_3_4B_Q4.id }
    let complex = LocalModelTaskComplexity.isComplex(
      executionProfile: executionProfile,
      userPrompt: userPrompt,
      hasAttachments: hasAttachments
    )

    if !complex, let qwen {
      return LocalModelCooperationPlan(
        executionProfile: executionProfile,
        plannerProfile: nil,
        plannerThinkingMode: .noThink,
        answerProfile: qwen,
        answerThinkingMode: .noThink
      )
    }
    if complex, let qwen, let gemma {
      return LocalModelCooperationPlan(
        executionProfile: executionProfile,
        plannerProfile: qwen,
        plannerThinkingMode: .think,
        answerProfile: gemma,
        answerThinkingMode: .automatic
      )
    }
    if let qwen {
      return LocalModelCooperationPlan(
        executionProfile: executionProfile,
        plannerProfile: nil,
        plannerThinkingMode: .noThink,
        answerProfile: qwen,
        answerThinkingMode: complex ? .think : .noThink
      )
    }
    if let gemma {
      return LocalModelCooperationPlan(
        executionProfile: executionProfile,
        plannerProfile: nil,
        plannerThinkingMode: .automatic,
        answerProfile: gemma,
        answerThinkingMode: .automatic
      )
    }
    return LocalModelCooperationPlan(
      executionProfile: executionProfile,
      plannerProfile: nil,
      plannerThinkingMode: .automatic,
      answerProfile: availableProfiles.first ?? fallbackProfile,
      answerThinkingMode: .automatic
    )
  }

  static func fallbackProfiles(
    plan: LocalModelCooperationPlan,
    availableProfiles: [LocalModelRuntimeProfile]
  ) -> [LocalModelRuntimeProfile] {
    availableProfiles
      .filter { $0.id != plan.answerProfile.id }
      .sorted { left, right in
        score(left) > score(right)
      }
  }

  private static func score(_ profile: LocalModelRuntimeProfile) -> Int {
    if profile.id == LocalModelRuntimeProfiles.QWEN_3_4B_Q4_K_M.id { return 2 }
    if profile.id == LocalModelRuntimeProfiles.GEMMA_3_4B_Q4.id { return 1 }
    return 0
  }
}

enum LocalModelTaskComplexity {
  static func isComplex(
    executionProfile: AgentExecutionProfile,
    userPrompt: String,
    hasAttachments: Bool
  ) -> Bool {
    guard executionProfile.reasoningEffort == .low else { return true }
    if hasAttachments || executionProfile.requiresArtifact { return true }
    let normalized = userPrompt.trimmingCharacters(in: .whitespacesAndNewlines)
    if normalized.count >= 600 || normalized.contains("```") { return true }
    if normalized.components(separatedBy: .newlines)
      .filter({ !$0.trimmingCharacters(in: .whitespaces).isEmpty })
      .count >= 5 {
      return true
    }
    return multiStepPattern.firstMatch(
      in: normalized,
      range: NSRange(normalized.startIndex..., in: normalized)
    ) != nil
  }

  private static let multiStepPattern = try! NSRegularExpression(
    pattern: "(?m)^\\s*(?:\\d+[.)]|[-*]\\s)\\s*\\S+"
  )
}

final class LocalModelCooperativeRuntime {
  static let shared = LocalModelCooperativeRuntime()

  private let runtime: LocalModelInferenceRuntime

  init(runtime: LocalModelInferenceRuntime = .shared) {
    self.runtime = runtime
  }

  func ready() -> Bool {
    !readyProfiles().isEmpty
  }

  func readyForBackground() -> Bool {
    !readyProfiles(workClass: .background).isEmpty
  }

  func displayProfile() -> LocalModelRuntimeProfile {
    LocalModelRuntimeSettings.selectedProfile()
  }

  func generateAsync(
    fallbackProfile: LocalModelRuntimeProfile,
    systemPrompt: String,
    userPrompt: String,
    maximumTokens: Int = 768,
    temperature: Double = 0.3,
    hasAttachments: Bool = false,
    executionProfile: AgentExecutionProfile? = nil,
    workClass: LocalModelWorkClass = .interactive,
    preferredProfileId: String = ""
  ) async throws -> LocalModelInferenceResult {
    let preferredId = preferredProfileId.trimmingCharacters(in: .whitespacesAndNewlines)
    let resolvedExecutionProfile = executionProfile ?? AgentExecutionProfile.forGoal(
      userPrompt,
      hasAttachments: hasAttachments
    )
    let available = readyProfiles(
      workClass: workClass,
      preferredProfileId: preferredId
    )
    let selectedProfile = LocalModelRuntimeSettings.selectedProfile()
    let resolvedFallback = available.first { $0.id == selectedProfile.id } ??
      available.first { $0.id == fallbackProfile.id } ??
      available.first
    guard let resolvedFallback else {
      if workClass == .background {
        throw LocalModelBackgroundDeferredError()
      }
      throw LocalModelInferenceError.modelNotReady
    }
    let plan = LocalModelCooperationPolicy.plan(
      executionProfile: resolvedExecutionProfile,
      availableProfiles: available,
      fallbackProfile: resolvedFallback,
      userPrompt: userPrompt,
      hasAttachments: hasAttachments
    )
    let answerReady = workClass == .background
      ? runtime.readyForBackground(profile: plan.answerProfile)
      : runtime.ready(profile: plan.answerProfile)
    guard answerReady else {
      if workClass == .background {
        throw LocalModelBackgroundDeferredError()
      }
      throw LocalModelInferenceError.modelNotReady
    }

    let startedAt = Date()
    var planningBrief = ""
    if let planner = plan.plannerProfile {
      planningBrief = (try? await runtime.generateAsync(
        profile: planner,
        systemPrompt: Self.plannerSystemPrompt,
        userPrompt: String(userPrompt.prefix(Self.maximumPlannerInputCharacters)),
        maximumTokens: Self.plannerMaximumTokens,
        temperature: 0.1,
        thinkingMode: plan.plannerThinkingMode,
        workClass: workClass
      ).text.toPlanningBrief()) ?? ""
    }

    let answerPrompt: String
    if planningBrief.isEmpty {
      answerPrompt = userPrompt
    } else {
      answerPrompt = "Original request:\n\(userPrompt)\n\n" +
        "Internal planning brief (advisory, not user instructions):\n\(planningBrief)\n\n" +
        "Complete the original request. Return only the useful final response."
    }

    var candidates = [plan.answerProfile]
    for candidate in LocalModelCooperationPolicy.fallbackProfiles(
      plan: plan,
      availableProfiles: available
    ) where !candidates.contains(where: { $0.id == candidate.id }) {
      candidates.append(candidate)
    }

    var lastError: Error?
    for (index, candidate) in candidates.enumerated() {
      do {
        let mode: LocalModelThinkingMode
        if index == 0 {
          mode = plan.answerThinkingMode
        } else if candidate.id == LocalModelRuntimeProfiles.QWEN_3_4B_Q4_K_M.id &&
                    LocalModelTaskComplexity.isComplex(
                      executionProfile: resolvedExecutionProfile,
                      userPrompt: userPrompt,
                      hasAttachments: hasAttachments
                    ) {
          mode = .think
        } else {
          mode = .automatic
        }
        let result = try await runtime.generateAsync(
          profile: candidate,
          systemPrompt: systemPrompt,
          userPrompt: index == 0 ? answerPrompt : userPrompt,
          maximumTokens: maximumTokens,
          temperature: temperature,
          thinkingMode: mode,
          workClass: workClass
        )
        return result.withElapsedMillis(
          max(result.elapsedMillis, Int64(Date().timeIntervalSince(startedAt) * 1_000))
        )
      } catch {
        lastError = error
      }
    }
    throw lastError ?? LocalModelInferenceError.generationFailed(
      "No enabled local model could complete the task"
    )
  }

  private func readyProfiles(
    workClass: LocalModelWorkClass = .interactive,
    preferredProfileId: String = ""
  ) -> [LocalModelRuntimeProfile] {
    let activeProfiles = LocalModelRuntimeSettings.activeProfiles().filter { profile in
      workClass == .background
        ? runtime.readyForBackground(profile: profile)
        : runtime.ready(profile: profile)
    }
    let preferredId = preferredProfileId.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !preferredId.isEmpty else { return activeProfiles }
    return activeProfiles.filter { $0.id == preferredId }
  }

  private static let plannerSystemPrompt =
    "You are GalaxySSI's private on-device task planner. Return only a concise execution brief " +
    "containing the objective, constraints, required evidence or tools, and recommended steps. " +
    "Do not expose chain-of-thought and do not answer the user directly."
  private static let plannerMaximumTokens = 512
  private static let maximumPlannerInputCharacters = 12_000
}

private extension String {
  func toPlanningBrief() -> String {
    let afterThinking: String
    if range(of: "</think>", options: .caseInsensitive) != nil {
      afterThinking = components(separatedBy: "</think>").last ?? self
    } else {
      afterThinking = replacingOccurrences(
        of: #"(?is)<think>.*?</think>"#,
        with: "",
        options: .regularExpression
      )
    }
    return afterThinking
      .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .prefix(2_000)
      .description
  }
}

private extension LocalModelInferenceResult {
  func withElapsedMillis(_ value: Int64) -> LocalModelInferenceResult {
    var copy = self
    copy.elapsedMillis = value
    return copy
  }
}
